sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dsl_scan.c

```diff
@@ -3433,6 +3446,14 @@ scan_exec_io(dsl_pool_t *dp, const blkptr_t *bp, int zio_flags,
 		mutex_exit(q_lock);
 	}
 
+	if (zio_flags & ZIO_FLAG_RESILVER)
+		scan_delay = zfs_resilver_delay;
+	else if (zio_flags & ZIO_FLAG_SCRUB)
```

Comment:
Is this check needed?  Can there be scan I/O which is neither scrub nor resilver?

Reply:
Yes, it's needed, as that determines which of the sysctl/tunbables it uses for the delay.  If there are other scan i/o types, they are not handled by the delay tunables.  (This is simply bringing back the old code, remember.)

Reply:
In old code I actually see one condition and one assertion. Are we looking on the same old code?

Reply:
Look at hash e161e8619f2f012952324e28f7604e2874c1bea4.  The original code was
        if (scn->scn_phys.scn_func == POOL_SCAN_SCRUB) {
                zio_flags |= ZIO_FLAG_SCRUB;
                zio_priority = ZIO_PRIORITY_SCRUB;
                needs_io = B_TRUE;
                scan_delay = zfs_scrub_delay;
        } else if (scn->scn_phys.scn_func == POOL_SCAN_RESILVER) {
                zio_flags |= ZIO_FLAG_RESILVER;
                zio_priority = ZIO_PRIORITY_RESILVER;
                needs_io = B_FALSE;
                scan_delay = zfs_resilver_delay;
        }

but the code went under major rewrite as part of the sorting change, and so I had to put it elsewhere.  Note that the tunables may in fact no longer be needed, but having them and defaulting to no delay doesn't cause any harm at the moment.

Reply:
You are looking too far back.  Before scrub rewrite it was:
        if (scn->scn_phys.scn_func == POOL_SCAN_SCRUB) {
                zio_flags |= ZIO_FLAG_SCRUB;
                needs_io = B_TRUE;
                scan_delay = zfs_scrub_delay;
        } else {
                ASSERT3U(scn->scn_phys.scn_func, ==, POOL_SCAN_RESILVER);
                zio_flags |= ZIO_FLAG_RESILVER;
                needs_io = B_FALSE;
                scan_delay = zfs_resilver_delay;
        }

Reply:
None of which changes the code.  Well, it doesn't cause an assert if it is a different kind of function, but since that code was gone in the rewrite, it's is not causing a change in behaviour from the rewrite.  (That is, my code uses zfs_resilver_delay if it's a resilver, and zfs_scrub_delay if it's a scrub, and no delay if it's neither one of those.)

Reply:
I won't insist on this if you don't like, but that check looks excessive to me.

Reply:
It *could* just be an else -- especially if combined with the old assertion -- but I wrote it that way because I thought it was clearer what it was doing and why.  There doesn't seem to be any advantage to having it be a simple if/else, though.

Reply:
If something should not happen, KASSERT() is generally a right way to check it.  Do as you wish.

Reply:
Are you happier with
        if (zio_flags & ZIO_FLAG_RESILVER)
                scan_delay = zfs_resilver_delay;
        else {
                ASSERT(zio_flags & ZIO_FLAG_SCRUB);
                scan_delay = zfs_scrub_delay;
        }

Reply:
Yes. That would closer represent existing code.

Reply:
Done then